### PR TITLE
Add additional WinAPI hook support and actions

### PR DESCRIPTION
**Pull request by ddash-ct**
* Added `shell32`, `shlwapi`, and `ws2_32` WinAPI base support
* Added to `kernel32` support
* Added `ShellOperation` action for usage with `ShellExecute` API

### CODE REVIEW

File: kordesii/utils/function_tracing/call_hooks/win_api/shell32.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    if dir_ptr:
+        directory = cpu_context.read_data(
+            dir_ptr, data_type=constants.WIDE_STRING if wide else constants.STRING
+        ).decode("utf-16-le" if wide else "utf8")
+    else:
+        directory = u""
+
+    logger.debug(f"{func_name}: {operation} {filepath}")
+    cpu_context.actions.append(
+        actions.ShellOperation(cpu_context.ip, operation, filepath, params, directory, visibility)
```
Comment (DC3-DCCI) [inline kordesii/utils/function_tracing/call_hooks/win_api/shell32.py:67]: To keep consistency and improve speed, please use logging format strings for log messages. 
For example, the following should be something like this:
```python
logger.debug("Writing temp directory %s to 0x%08x", temp_dir, buffer_ptr)
```

This applies to all logging message.

File: kordesii/utils/function_tracing/call_hooks/win_api/shell32.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        operation = u""
+
+    if dir_ptr:
+        directory = cpu_context.read_data(
+            dir_ptr, data_type=constants.WIDE_STRING if wide else constants.STRING
+        ).decode("utf-16-le" if wide else "utf8")
+    else:
+        directory = u""
+
+    logger.debug(f"{func_name}: {operation} {filepath}")
```
Comment (DC3-DCCI) [inline kordesii/utils/function_tracing/call_hooks/win_api/shell32.py:65]: To keep consistency and improve speed, please use logging format strings for log messages. 
For example, the following should be something like this:
```python
logger.debug("Writing temp directory %s to 0x%08x", temp_dir, buffer_ptr)
```
This applies to all logging message

File: kordesii/utils/function_tracing/call_hooks/win_api/shell32.py (hunk 1 above)
Comment (DC3-DCCI) [inline kordesii/utils/function_tracing/call_hooks/win_api/shell32.py:67]: Should the `wc.Visibility` enum be used? See the "WinExec" hook in kernel32.py for an example of its usage.

File: kordesii/utils/function_tracing/call_hooks/win_api/shell32.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    _, path_ptr, csidl, _ = func_args
+    csidl_int = construct.Int32ul.build(csidl)
+    try:
+        folder = construct.KnownFolderID(construct.Int32ul).parse(csidl_int)
+    except construct.ConstructError:
+        logger.warning(f"Could not acquire a folder for CSIDL {csidl}")
+        folder = f"{csidl}"
+    logger.debug(f"Writing CSIDL {folder} to 0x{path_ptr:08x}")
+    cpu_context.write_data(path_ptr, folder, data_type=constants.WIDE_STRING if wide else constants.STRING)
+    return True
```
Comment (DC3-DCCI) [inline kordesii/utils/function_tracing/call_hooks/win_api/shell32.py:114]: This is an invalid data type returned and will be ignored by the emulator.
Function hooks should either return an integer, which will be set in rax, or None in order to not affect the rax register.

Should `wc.ERROR_SUCCESS` be returned instead?

File: kordesii/utils/function_tracing/call_hooks/win_api/shlwapi.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    curr_path = cpu_context.read_data(
+        path_ptr, data_type=constants.WIDE_STRING if wide else constants.STRING
+    ).decode("utf-16-le" if wide else "utf8")
+    more_path = cpu_context.read_data(
+        more_ptr, data_type=constants.WIDE_STRING if wide else constants.STRING
+    ).decode("utf-16-le" if wide else "utf8")
+
+    full_path = ntpath.join(curr_path, more_path)
+    cpu_context.write_data(path_ptr, full_path, data_type=constants.WIDE_STRING if wide else constants.STRING)
+    return True
```
Comment (DC3-DCCI) [inline kordesii/utils/function_tracing/call_hooks/win_api/shlwapi.py:45]: See previous comment about returning integers.

File: kordesii/utils/function_tracing/call_hooks/win_api/shell32.py (hunk 3 above)
Comment (ddash-ct) [reply kordesii/utils/function_tracing/call_hooks/win_api/shell32.py:114]: `SHGetSpecialFolderPath` explicitly returns a `BOOL` value, defined in [MSDN](https://docs.microsoft.com/en-us/windows/win32/api/shlobj_core/nf-shlobj_core-shgetspecialfolderpatha) as `True` or `False`, whereas `wc.ERROR_SUCCESS` is `0`, which would imply `False`. I can adjust this to return `1`, or can add to the `win_constants.py` to add `TRUE = 1` and `FALSE = 0`?

Comment (DC3-DCCI) [reply kordesii/utils/function_tracing/call_hooks/win_api/shell32.py:114]: The later option would be preferred.  Alternatively, you could update the `call` opcode to convert returned `True` and `False` values into `1` and `0` respectively.

File: kordesii/utils/function_tracing/call_hooks/win_api/shell32.py (hunk 2 above)
Comment (ddash-ct) [reply kordesii/utils/function_tracing/call_hooks/win_api/shell32.py]: Addressed in 9862ed7

File: kordesii/utils/function_tracing/call_hooks/win_api/shell32.py (hunk 1 above)
Comment (ddash-ct) [reply kordesii/utils/function_tracing/call_hooks/win_api/shell32.py]: Updated to use the `wc.Visibility` enum in 5f2c519

File: kordesii/utils/function_tracing/call_hooks/win_api/shell32.py (hunk 3 above)
Comment (ddash-ct) [reply kordesii/utils/function_tracing/call_hooks/win_api/shell32.py:114]: Updated the `call` opcode in 5f2c519 to convert the returned `True` and `False` values as described

File: kordesii/utils/function_tracing/call_hooks/win_api/shlwapi.py (hunk 4 above)
Comment (ddash-ct) [reply kordesii/utils/function_tracing/call_hooks/win_api/shlwapi.py:45]: Updated the `call` opcode in 5f2c519 to convert the returned `True` and `False` values as described

**Pull request merged**